### PR TITLE
refactor: use new HikesViewModel

### PR DESCRIPTION
**Pull request by MaelImhof**
# Summary

This PR replaces the previous `ListOfHikeRoutesViewModel` and `SavedHikesViewModel` with the new `HikesViewModel` in the `MapScreen`, `SavedHikesScreen` and `DetailsScreen`.

This replacement makes the code more readable, maintanable but also makes the app faster and more reliable, in particular regarding the elevation graphs.

# Details

- `MapScreen` and its tests were adapted to use `HikesViewModel`
- `SavedHikesScreen` and its tests were adapted to use `HikesViewModel`
- `DetailsScreen` and its tests were adapted to use `HikesViewModel`
- `RunHikeScreen` is left untouched for now, to allow for working on it in parallel. `HikesViewModel` will be added in later (but soon)

As a natural addition, since it became very simple, the following features now work :

- Elevation graphs are displayed in the saved hikes screen
- Clicking on a hike in the saved hikes screen will open the details screen for that hike

There are some known bugs that need to be addressed but that I considered out-of-scope for this PR:

- If you open the details screen from the saved hikes screen (click on a saved hike) and unsave it, it will close the details screen and go back to the saved hikes screen. This is expected because when unsaving the hike, it effectively unloads it by design. The `HikesViewModel` itself will need to be updated to fix it.
- Error handling is poor for now, because another task exists that consists in adding better error handling to this view model.
- If an error occurs while loading the details of a hike (for example its elevation), nothing tells it to the view except the callback, so you might get stuck on the loading screen. This is related to the poor error handling. It is also related to how the elevation service works, there is another task to fix that as well.

Anyway, the app does not work flawlessly and perfectly yet, but it is way more stable, usable and fast. Let me know what you think !

### CODE REVIEW

File: app/src/main/java/ch/hikemate/app/ui/map/HikeDetailsScreen.kt
Comment (rBergCode) [inline app/src/main/java/ch/hikemate/app/ui/map/HikeDetailsScreen.kt:271]: Nitpick: I think you could put `hike.color` directly as the parameter. This is longer 😂

File: app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt
Comment (rBergCode) [inline app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt:73]: Question: Why don't you use the `hikesType` here ?

File: app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt
Comment (rBergCode) [inline app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt:63]: Warning: You have a sonar issue here. Please correct it

File: app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt
Comment (rBergCode) [inline app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt:186]: Nitpick: Can you refactor this part to not have the same data computed twice in the code (for example `title`)

File: app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt
Comment (rBergCode) [inline app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt:228]: Same here

File: app/src/androidTest/java/ch/hikemate/app/ui/map/MapScreenTest.kt
Comment (rBergCode) [inline app/src/androidTest/java/ch/hikemate/app/ui/map/MapScreenTest.kt:343]: Nitpick: Can you remove that please

File: app/src/androidTest/java/ch/hikemate/app/ui/saved/SavedHikesScreenTest.kt
Comment (rBergCode) [inline app/src/androidTest/java/ch/hikemate/app/ui/saved/SavedHikesScreenTest.kt:214]: Question: Why did you remove that ?

**Review by rBergCode — CHANGES_REQUESTED** [review]
What a change! This is very nice and makes the app run much smoother.
I have some suggestions before you can merge it.

File: app/src/main/java/ch/hikemate/app/utils/MapUtils.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/utils/MapUtils.kt:42]: There were two reasons behind these changes:

- `showHikeOnMap` does not need anything else than the waypoints, so the idea was to only request what is required
- We could pass the hike, but now the waypoints of the hike are of type `DeferredData`, so it would be a pain to handle this in `showHikeOnMap`. It's easier to handle it before and once we have them call the function.

File: app/src/main/java/ch/hikemate/app/ui/components/CenteredLoadingAnimation.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/components/CenteredLoadingAnimation.kt:31]: I'm not sure how it could be more explicit, do you have an idea in mind?

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:701]: Done 👌🏻

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:706]: Done 👌🏻

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:724]: Very good catch, done 👌🏻

File: app/src/main/java/ch/hikemate/app/ui/map/HikeDetailsScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/HikeDetailsScreen.kt:138]: Fixed 👌🏻

File: app/src/main/java/ch/hikemate/app/utils/MapUtils.kt
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/utils/MapUtils.kt:42]: Good second point. Then we can keep it like this.

File: app/src/main/java/ch/hikemate/app/ui/components/CenteredLoadingAnimation.kt
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/ui/components/CenteredLoadingAnimation.kt:31]: `textUnderLoadingAnimation` ?😂

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:638]: Wouldn't that be better to either
- Remove the loaded hikes when switching a tab
- Remove the loaded hikes when going to the map screen

How is this handled from map -> saved hikes ?

File: app/src/main/java/ch/hikemate/app/ui/components/CenteredLoadingAnimation.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/components/CenteredLoadingAnimation.kt:31]: Well technically the text can be either above or under the loading animation, based on the second parameter. The second parameter is a boolean that defines whether the text should be put above the loading animation or not (if not, the text will go under the loading animation)?

Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/components/CenteredLoadingAnimation.kt:31]: Maybe I can rename `textAboveLoadingAnimation` to `putTextAboveLoadingAnimation`, or have two parameters `textAboveLoadingAnimation` and `textUnderLoadingAnimation`.

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:638]: If we remove the hikes, we need to re-compute them, which is not the case if we keep them without displaying.

When you go back to the map, the hikes list is not cleared, it contains the saved hikes. Once you search for hikes, the list will be cleared, except for the hikes that are already loaded. Those ones will be kept and already have all of their data.

When going from map screen to saved hikes, the same happens, but the list is cleared right away because it loads saved hikes right away.

I could add a function to the view model to clear the hikes list, but I'm not sure it's necessary.

File: app/src/main/java/ch/hikemate/app/ui/map/HikeDetailsScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/HikeDetailsScreen.kt:187]: Thank you 🥳

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:677]: I refactored it to avoid code duplication.

File: app/src/main/java/ch/hikemate/app/ui/map/HikeDetailsScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/HikeDetailsScreen.kt:271]: Why did I do this in the first place 😂

File: app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt:73]: No excuse your honor, I was only using `loading`, but if there's an error it would be better to use `hikesType`

File: app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt:63]: image

This is a personal best

File: app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt:186]: Done 👌🏻

File: app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt:228]: Done 👌🏻

File: app/src/androidTest/java/ch/hikemate/app/ui/map/MapScreenTest.kt
Comment (MaelImhof) [reply app/src/androidTest/java/ch/hikemate/app/ui/map/MapScreenTest.kt:343]: Done, sorry about that 😂

File: app/src/androidTest/java/ch/hikemate/app/ui/saved/SavedHikesScreenTest.kt
Comment (MaelImhof) [reply app/src/androidTest/java/ch/hikemate/app/ui/saved/SavedHikesScreenTest.kt:214]: There is no error handling for now in the view model, so nothing happens if an error occurs in the repository, there is nothing to test, except that the app stays on the loading screen forever...

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:640]: Thanks for the explanation. It's fine like this

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt
Comment (rBergCode) [inline app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:722]: Suggestion: Wouldn't that be better to change `suitableLabelIcon` to `messageIcon` and do `painterResource(R.drawable.{case})` in both case to reduce this weird condition here ?

File: app/src/androidTest/java/ch/hikemate/app/ui/saved/SavedHikesScreenTest.kt
Comment (rBergCode) [reply app/src/androidTest/java/ch/hikemate/app/ui/saved/SavedHikesScreenTest.kt:214]: Okay, I've seen there is a task for this in the PB. Sounds good

**Review by rBergCode — CHANGES_REQUESTED** [review]
Thanks for the changes! I have a small suggestion more and then we're good to go!

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:722]: Yup, that's better indeed

**Comment by MaelImhof** [comment]
CI won't pass because of the ongoing issue with the elevation API, but the PR should be good to go. I might want to wait for #237 to be merged so that I can solve any conflicts.

**Review by rBergCode — DISMISSED** [review]
Thanks for the last change!

**Review by rBergCode — APPROVED** [review]
Still all good!

**Comment by MaelImhof** [comment]
Woups sorry about that

**Review by rBergCode — APPROVED** [review]
Oops

**Pull request merged**
